zerver/tests/test_link_embed.py

```diff
@@ -240,6 +241,55 @@ def _send_message_with_test_org_url(self, sender_email, queue_should_run=True, r
         msg = Message.objects.select_related("sender").get(id=msg_id)
         return msg
 
+    @override_settings(INLINE_URL_EMBED_PREVIEW=True)
+    def test_message_update_race_condition(self):
+        # type: () -> None
+        email = self.example_email('hamlet')
+        self.login(email)
+        original_url = 'http://test.org/'
+        edited_url = 'http://edited.org/'
+        with mock.patch('zerver.lib.actions.queue_json_publish') as patched:
+            msg_id = self.send_stream_message(email, "Scotland",
+                                              topic_name="foo", content=original_url)
+            patched.assert_called_once()
+            queue = patched.call_args[0][0]
+            self.assertEqual(queue, "embed_links")
+            event = patched.call_args[0][1]
+
+        def wrapped_queue_json_publish(*args, **kwargs):
+            # type: (*Any, **Any) -> None
+            # Mock the network request result so the test can be fast without Internet
+            response = MockPythonResponse(self.open_graph_html, 200)
+            mocked_response_original = mock.Mock(
+                side_effect=lambda k: {original_url: response}.get(k, MockPythonResponse('', 404)))
+            mocked_response_edited = mock.Mock(
+                side_effect=lambda k: {edited_url: response}.get(k, MockPythonResponse('', 404)))
+            with self.settings(TEST_SUITE=False, CACHES=TEST_CACHES):
+                with mock.patch('requests.get', mocked_response_original):
+                    # Run the queue processor. This will simulate the event for original_url being
+                    # processed after the message has been edited.
+                    FetchLinksEmbedData().consume(event)
+            msg = Message.objects.select_related("sender").get(id=msg_id)
+            # The content of the message has changed since the event for original_url has been created,
+            # it should not be rendered. Another, up-to-date event will have been sent (edited_url).
+            self.assertNotIn('<a href="{0}" target="_blank" title="The Rock">The Rock</a>'.format(original_url),
+                             msg.rendered_content)
```

Comment:
I think maybe a `mocked_response_edited.assert_not_called()` sort of thing would be useful here?